# **Tugas 2**


Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

- Menggunakan data spam.csv

- Fitur CountVectorizer dengan mengaktifkan stop_words

- Evaluasi hasilnya

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

- Menggunakan data spam.csv

- Fitur TF-IDF dengan mengaktifkan stop_words

- Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

- Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [ ]:
import pandas as pd

# 1. Load the 'spam.csv' dataset
df = pd.read_csv('/content/drive/MyDrive/spam.csv', encoding='latin-1')

# 2. Drop the last three columns
df = df.drop(df.iloc[:,2:], axis=1)

# 3. Rename the remaining columns
df.columns = ['Labels', 'Messages']

# 4. Inspect the preprocessed DataFrame
print("DataFrame Info:")
df.info()
print("\nDescriptive Statistics:")
print(df.describe())
print("\nLabel Value Counts:")
print(df['Labels'].value_counts())

# 5. Encode the 'Labels' column
new_labels = {
    'spam': 1,
    'ham': 0
}
df['Labels'] = df['Labels'].map(new_labels)

# 6. Separate features (X) and label (y)
X = df['Messages'].values
y = df['Labels'].values

print("\nDataFrame after Label Encoding (first 5 rows):")
print(df.head())
print(f"\nX shape: {X.shape}")
print(f"y shape: {y.shape}")

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Labels    5572 non-null   object
 1   Messages  5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB

Descriptive Statistics:
       Labels                Messages
count    5572                    5572
unique      2                    5169
top       ham  Sorry, I'll call later
freq     4825                      30

Label Value Counts:
Labels
ham     4825
spam     747
Name: count, dtype: int64

DataFrame after Label Encoding (first 5 rows):
   Labels                                           Messages
0       0  Go until jurong point, crazy.. Available only ...
1       0                      Ok lar... Joking wif u oni...
2       1  Free entry in 2 a wkly comp to win FA Cup fina...
3       0  U dun say so early hor... U c already then say...
4       0  Nah I don't think he goes 

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# 1. Split the X and y data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 2. Initialize a CountVectorizer object with stop_words='english'
bow_cv = CountVectorizer(stop_words='english')

# 3. Apply fit_transform to X_train and transform to X_test
X_train_cv = bow_cv.fit_transform(X_train)
X_test_cv = bow_cv.transform(X_test)

print(f"Shape of X_train_cv: {X_train_cv.shape}")
print(f"Shape of X_test_cv: {X_test_cv.shape}")

# 4. Initialize a MultinomialNB model
mnb_cv = MultinomialNB()

# 5. Fit the MultinomialNB model to X_train_cv and y_train
mnb_cv.fit(X_train_cv, y_train)

# 6. Make predictions on X_test_cv
y_pred_cv = mnb_cv.predict(X_test_cv)

# 7. Calculate and print the evaluation metrics
print("\n--- Evaluation with CountVectorizer (stop_words='english') ---")
print(f"Accuracy: {accuracy_score(y_test, y_pred_cv):.4f}")
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_cv))
print("\nClassification Report:\n", classification_report(y_test, y_pred_cv))

Shape of X_train_cv: (4457, 7472)
Shape of X_test_cv: (1115, 7472)

--- Evaluation with CountVectorizer (stop_words='english') ---
Accuracy: 0.9839

Confusion Matrix:
 [[959   6]
 [ 12 138]]

Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.99      0.99       965
           1       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# 1. Split the X and y data into training and testing sets (re-splitting to ensure consistency)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 2. Initialize a TfidfVectorizer object with stop_words='english'
tfidf = TfidfVectorizer(stop_words='english')

# 3. Apply fit_transform to X_train and transform to X_test
X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

print(f"Shape of X_train_tfidf: {X_train_tfidf.shape}")
print(f"Shape of X_test_tfidf: {X_test_tfidf.shape}")

# 4. Initialize a MultinomialNB model
mnb_tfidf = MultinomialNB()

# 5. Fit the MultinomialNB model to X_train_tfidf and y_train
mnb_tfidf.fit(X_train_tfidf, y_train)

# 6. Make predictions on X_test_tfidf
y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)

# 7. Calculate and print the evaluation metrics
print("\n--- Evaluation with TfidfVectorizer (stop_words='english') ---")
print(f"Accuracy: {accuracy_score(y_test, y_pred_tfidf):.4f}")
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_tfidf))
print("\nClassification Report:\n", classification_report(y_test, y_pred_tfidf))

Shape of X_train_tfidf: (4457, 7472)
Shape of X_test_tfidf: (1115, 7472)

--- Evaluation with TfidfVectorizer (stop_words='english') ---
Accuracy: 0.9668

Confusion Matrix:
 [[965   0]
 [ 37 113]]

Classification Report:
               precision    recall  f1-score   support

           0       0.96      1.00      0.98       965
           1       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115



Perbandingan

1. Naive Bayes + CountVectorizer

  - Akurasi: 98.39%

  - F1-score spam: 0.94

  - Kelebihan: Seimbang antara precision (0.96) dan recall (0.92). Deteksi spam lebih stabil.

2. Naive Bayes + TfidfVectorizer

  - Akurasi: 96.68%

  - F1-score spam: 0.86

  - Kelemahan: Meskipun precision sempurna (1.00), tapi recall rendah (0.75) — banyak spam yang terlewat.

Kesimpulan

Model Naive Bayes dicampur dengan CountVectorizer lebih unggul karena akurasi dan kemampuan untuk mendeteksi spam lebih baik. CountVectorizer lebih efektif menangkap pola kata mentah pada dataset spam.csv dibanding TF-IDF.